wen-restart/src/last_voted_fork_slots_aggregate.rs

```diff
@@ -0,0 +1,245 @@
+use {
+    crate::solana::wen_restart_proto::{
+        LastVotedForkSlotsAggregateRecord, LastVotedForkSlotsRecord,
+    },
+    log::*,
+    solana_gossip::restart_crds_values::RestartLastVotedForkSlots,
+    solana_runtime::epoch_stakes::EpochStakes,
+    solana_sdk::{clock::Slot, pubkey::Pubkey},
+    std::collections::{HashMap, HashSet},
+};
+
+pub struct LastVotedForkSlotsAggregate {
+    root_slot: Slot,
+    repair_threshold: f64,
+    // TODO(wen): using local root's EpochStakes, need to fix if crossing Epoch boundary.
+    epoch_stakes: EpochStakes,
+    last_voted_fork_slots: HashMap<Pubkey, RestartLastVotedForkSlots>,
+    slots_stake_map: HashMap<Slot, u64>,
+    active_peers: HashSet<Pubkey>,
+    slots_to_repair: HashSet<Slot>,
+}
+
+pub struct LastVotedForkSlotsAggregateResult {
+    pub slots_to_repair: Vec<Slot>,
+    pub active_percent: f64, /* 0 ~ 100.0 */
+}
+
+impl LastVotedForkSlotsAggregate {
+    pub(crate) fn new(
+        root_slot: Slot,
+        repair_threshold: f64,
+        epoch_stakes: &EpochStakes,
+        last_voted_fork_slots: &Vec<Slot>,
+        my_pubkey: &Pubkey,
+    ) -> Self {
+        let mut active_peers = HashSet::new();
+        let my_stake = Self::validator_stake(epoch_stakes, my_pubkey);
+        active_peers.insert(*my_pubkey);
+        let mut slots_stake_map = HashMap::new();
+        for slot in last_voted_fork_slots {
+            if slot <= &root_slot {
+                break;
+            }
+            slots_stake_map.insert(*slot, my_stake);
+        }
+        Self {
+            root_slot,
+            repair_threshold,
+            epoch_stakes: epoch_stakes.clone(),
+            last_voted_fork_slots: HashMap::new(),
+            slots_stake_map,
+            active_peers,
+            slots_to_repair: HashSet::new(),
+        }
+    }
+
+    fn validator_stake(epoch_stakes: &EpochStakes, pubkey: &Pubkey) -> u64 {
+        epoch_stakes
+            .node_id_to_vote_accounts()
+            .get(pubkey)
+            .map(|x| x.total_stake)
+            .unwrap_or_default()
+    }
+
+    pub(crate) fn aggregate(
+        &mut self,
+        new_last_voted_fork_slots: Vec<RestartLastVotedForkSlots>,
```

Comment:
For unit testing, I think it's cleaner to take a single `RestartLastVotedForkSlots` and return an additional `Option<LastVotedForkSlotsRecord>` depending if stake is 0. 

Then we don't need to pass the `last_voted_fork_slots_aggregate_record` argument, the caller can insert the `LastVotedForkSlotsRecord` themselves. This keeps this function's purpose and unit tests cleaner

Reply:
Done.